File: frontend/src/api/types/tournaments.ts
export type PokerApp = "pppoker" | "xpoker" | "poker21" | "other";
export type BountyKind = "none" | "ko" | "pko" | "mystery";
export type GameType = "nlh" | "plo" | "plo5" | "mixed" | "other";
export type DayPeriod = "day" | "evening" | "night";
export type ScheduleView = "cards" | "table";

export interface TournamentClub {
  id: string;
  name: string;
  slug: string;
  app: PokerApp;
  /** Курс: 1 фишка = chip_value в chip_currency_code. Суммы турнира — в фишках. */
  chip_value: string | null;
  chip_currency_code: string | null;
  /** Символ перед суммой; USDT приходит как «$». */
  currency_symbol: string | null;
}

export interface Tournament {
  id: string;
  name: string;
  game_type: GameType;
  bounty_kind: BountyKind;
  buyin: string;
  guarantee: string | null;
  rebuy_cost: string | null;
  rebuy_terms: string | null;
  addon_cost: string | null;
  addon_terms: string | null;
  start_stack: number | null;
  table_size: number | null;
  late_reg_levels: number | null;
  level_minutes: string | null;
  structure: string | null;
  ticket_value: string | null;
  /** Куда ведёт сателлит: показываем «Sat → цель». */
  satellite_target: string | null;
  early_bird_players: number | null;
  /** Имя в лобби приложения — игроку показываем его; имя с афиши остаётся только в базе. */
  lobby_name: string | null;
  /** Какая часть бай-ина уходит в баунти, %. */
  bounty_share: number | null;
  /** Бонус Early Bird: «+50% фишек» — тем, кто сел до конца early_bird_levels уровней. */
  early_bird_bonus: string | null;
  early_bird_levels: number | null;
  /** Путь в живую серию (X-Poker): событие, даты, шаг; шаг пустой — сам турнир серии. */
  live_event?: string | null;
  live_dates?: string | null;
  live_step?: number | null;
  has_jackpot: boolean;
  notes: string | null;
  club: TournamentClub;
  starts_at: string;
  late_reg_closes_at: string | null;
  status: "scheduled" | "cancelled";
  is_promoted: boolean;
  buyin_rub: string | null;
  guarantee_rub: string | null;
  has_addon: boolean;
  /** До какого момента ещё можно получить бонус Early Bird. */
  early_bird_closes_at?: string | null;
}

/** Колокольчик: за 5 минут до старта или до конца поздней регистрации (ответ 11.7). */
/** Живая серия: шаги пути и сам турнир серии, отдельно от онлайн-расписания. */
export interface LiveEvent {
  title: string;
  dates: string | null;
  club: TournamentClub;
  items: Tournament[];
}

export type ReminderKind = "start" | "late_reg";

export interface TournamentReminder {
  tournament_id: string;
  kind: ReminderKind;
}

export interface TournamentsParams {
  from?: string;
  to?: string;
  app?: PokerApp[];
  period?: DayPeriod[];
  buyin_rub_min?: number;
  buyin_rub_max?: number;
}

File: backend/app/schemas/tournaments.py
from __future__ import annotations

from datetime import date, datetime, time
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import BountyKind, GameType, PokerApp, ReminderKind, TournamentStatus


class TournamentFields(BaseModel):
    """Поля турнира, общие для шаблона и старта. Суммы — в фишках клуба."""

    name: str = Field(min_length=1, max_length=160)
    game_type: GameType = GameType.NLH
    bounty_kind: BountyKind = BountyKind.NONE
    buyin: Decimal = Field(ge=0)
    guarantee: Decimal | None = Field(default=None, ge=0)
    rebuy_cost: Decimal | None = Field(default=None, ge=0)
    rebuy_terms: str | None = Field(default=None, max_length=32)
    addon_cost: Decimal | None = Field(default=None, ge=0)
    addon_terms: str | None = Field(default=None, max_length=32)
    start_stack: int | None = Field(default=None, ge=0)
    table_size: int | None = Field(default=None, ge=2, le=10)
    late_reg_levels: int | None = Field(default=None, ge=1)
    level_minutes: str | None = Field(default=None, max_length=16)
    structure: str | None = Field(default=None, max_length=32)
    ticket_value: Decimal | None = Field(default=None, ge=0)
    satellite_target: str | None = Field(default=None, max_length=160)
    early_bird_players: int | None = Field(default=None, ge=0)
    lobby_name: str | None = Field(default=None, max_length=160)
    bounty_share: int | None = Field(default=None, ge=1, le=100)
    early_bird_bonus: str | None = Field(default=None, max_length=64)
    early_bird_levels: int | None = Field(default=None, ge=1, le=50)
    has_jackpot: bool = False
    live_event: str | None = Field(default=None, max_length=160)
    live_dates: str | None = Field(default=None, max_length=64)
    live_step: int | None = Field(default=None, ge=1, le=5)
    notes: str | None = None


class TemplateDraft(TournamentFields):
    """Шаблон регулярного турнира, извлечённый из источника (CSV союза, афиша)."""

    weekdays: list[int] = Field(min_length=1)
    start_time: time
    # Разовое событие (турнир месяца, серия): valid_from = valid_until = дата.
    valid_from: date | None = None
    valid_until: date | None = None
    # Турнир месяца: номер недели месяца (1–5) или -1 — последняя.
    month_week: int | None = Field(default=None, ge=-1, le=5)
    late_reg_close_offset_min: int | None = Field(default=None, ge=0)


class ParseIssue(BaseModel):
    row: int | None = None
    message: str


class TemplateParseResult(BaseModel):
    templates: list[TemplateDraft]
    issues: list[ParseIssue] = Field(default_factory=list)
    rows_total: int = 0


class TemplateRead(TemplateDraft):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    club_id: UUID
    is_active: bool
    source: str | None
    is_promoted: bool


class TemplatesImportResult(BaseModel):
    """Итог импорта сетки. При dry_run изменения посчитаны и откатаны."""

    dry_run: bool
    rows_total: int
    templates_parsed: int
    issues: list[ParseIssue]
    templates_created: int
    templates_updated: int
    templates_unchanged: int
    templates_removed: int
    tournaments_created: int
    tournaments_updated: int
    tournaments_deleted: int
    tournaments_detached_kept: int


class TournamentClub(BaseModel):
    id: UUID
    name: str
    slug: str
    app: PokerApp
    chip_value: Decimal | None
    chip_currency_code: str | None
    # Символ перед суммой: USDT показываем как «$» (решение Ивана 2026-09-13).
    currency_symbol: str | None
    # ID клуба в приложении — для приложений без диплинков игрок входит по нему.
    app_club_id: str | None = None


class TournamentRead(TournamentFields):
    id: UUID
    club: TournamentClub
    starts_at: datetime
    late_reg_closes_at: datetime | None
    status: TournamentStatus
    is_promoted: bool
    # Бай-ин в рублях по курсу клуба — «примерно», для фильтра и подписи. None, если курс
    # клуба или валюты не задан.
    buyin_rub: Decimal | None
    guarantee_rub: Decimal | None
    has_addon: bool
    # До какого момента действует Early Bird: старт + N уровней с перерывами, как у поздней реги.
    early_bird_closes_at: datetime | None = None
    app_link: str | None = None


class TemplateDeleteResult(BaseModel):
    """Турнир убран из сетки: сколько будущих стартов пропало из расписания."""

    tournaments_deleted: int


class TournamentReminderRead(BaseModel):
    tournament_id: UUID
    kind: ReminderKind


class TournamentRemindersUpdate(BaseModel):
    """Какие колокольчики должны стоять на турнире; пустой список снимает все."""

    kinds: list[ReminderKind] = Field(default_factory=list, max_length=2)


class LiveEventRead(BaseModel):
    """Живая серия: шаги-сателлиты и сам турнир серии, отдельно от онлайн-расписания."""

    title: str
    dates: str | None
    club: TournamentClub
    items: list[TournamentRead]
